File: plantuml_to_drawio_xml.pl
#!/usr/bin/perl

use strict;
use MIME::Base64;
use Time::gmtime;
use File::stat;
# use open qw( :encoding(UTF-8) :std ); # Set UTF-8 as the default encoding
use autodie;
use IPC::Open2;

BEGIN {
  use vars qw($JAVABIN $PLANTUMLJAR $DEBUG);
  $JAVABIN="java";
  $PLANTUMLJAR="plantuml-1.2024.4.jar";
  $DEBUG=0;
  use vars qw($DIO_AGENT_NAME $DIO_ETAG_ID $DIO_DIAGRAM_ID $DIO_USEROBJECT_ID);
  $DIO_AGENT_NAME    = "5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.0.0 Safari/537.36";
  $DIO_ETAG_ID       = "ETAG111111111111xx01";
  $DIO_DIAGRAM_ID    = "DIAGRAM111111111xx01";
  $DIO_USEROBJECT_ID = "USEROBJECT111111xx01-1";
}

sub dprint (@) {
  print @_ if $DEBUG >= 1;
}

my $plantuml_file = shift;
my $cmd_err=0;
unless ( -f "$plantuml_file" ) {
  print "Usage: $0 <PlantUML-file>\n";
  $cmd_err++;
}
unless ( -f "$PLANTUMLJAR" ) {
  print "The java binary should be in the current path, and PlantUML jar in the current directory.\n";
  $cmd_err++;
}
exit 0 if $cmd_err;
my $i=1;

#
# Convert PlantUML plain text data to SVG data
#
# Read in PlantUML
my $plantuml_textdata;
my $plantuml_svgdata;
open (TPUML,"<$plantuml_file");
while (<TPUML>) {
  $plantuml_textdata .= "$_";
}
chomp($plantuml_textdata);
dprint ($i++,"PlantUML_Text=[$plantuml_textdata]\n");
close (TPUML);
#
# Convert PlantUML to SVG using PlanUML Java Library
my ($stdout,$stdin);
my $pid = open2($stdout, $stdin, "$JAVABIN -jar $PLANTUMLJAR -tsvg -pipe");
print $stdin "$plantuml_textdata\n";
close($stdin);
while (my $line = <$stdout>) {
  $plantuml_svgdata .= "$line";
  #last if "$line" =~ /\<\/svg\>/;
}
close($stdout);
waitpid $pid, 0;
dprint ($i++,"PlantUML_SVG=[$plantuml_svgdata]\n");

#
# base64 encode the PlantUML SVG data
#
my $plantuml_svgb64data = encode_base64($plantuml_svgdata, '');
dprint ($i++,"PlantUML_Base64=[$plantuml_svgb64data]\n");

#
# XML/HTML/URL encode the PlantUML text data
# Only 5 characters are needed to be escaped for XML, and include newlines and tabs
#
my $plantuml_escapedtextdata = $plantuml_textdata;
$plantuml_escapedtextdata =~ s/\&/\&amp;/g;
$plantuml_escapedtextdata =~ s/</\&lt;/g;
$plantuml_escapedtextdata =~ s/>/\&gt;/g;
$plantuml_escapedtextdata =~ s/"/\&quot;/g;
$plantuml_escapedtextdata =~ s/'"'"'/\&apos;/g;
$plantuml_escapedtextdata =~ s/\r\n/\\n/g;
$plantuml_escapedtextdata =~ s/\n/\\n/g;
$plantuml_escapedtextdata =~ s/\r/\\n/g; #lost carriage returns
$plantuml_escapedtextdata =~ s/\t/\\t/g;
dprint ($i++,"PlantUML_EscapedText=[$plantuml_escapedtextdata]\n");

#
# Obtain the SVG dimensions
#
my ($width,$height) = $plantuml_svgdata =~ /.*width:([\d]+px);height:([\d]+px);.*/;
dprint ($i++,"[width=$width,height=$height]\n");

#
# Set the Draw.io revision date to the last modified date of the PlantUML file
#
my $filegmtdate = gmtime(stat($plantuml_file)->mtime);
dprint ($i++,"[gmtime=$filegmtdate]\n");
my $year=($filegmtdate->year() + 1900);
my $month=sprintf("%02d",($filegmtdate->mon() +1));
my $mday=sprintf("%02d",$filegmtdate->mday());
my $hour=sprintf("%02d",$filegmtdate->hour());
my $min=sprintf("%02d",$filegmtdate->min());
my $sec=sprintf("%02d",$filegmtdate->sec());
my $gmt_timestamp = (join("-",($year,$month,$mday)) . "T" . join(":",($hour,$min,$sec)) . ".000Z"); # i.e. 2022-06-17T16:28:54.000Z
dprint ($i++,"[gmt_timestamp=$gmt_timestamp]\n");

#
# Create the Draw.io file with imported PlantUML code and SVG
# XML Data Template pulled from a random source file
#
my $drawio_xmldata = ('<?xml version="1.0" encoding="UTF-8"?>
<mxfile host="app.diagrams.net" modified="' . $gmt_timestamp . '" agent="' . $DIO_AGENT_NAME . '" etag="' . $DIO_ETAG_ID . '" version="20.0.1" type="embed">
  <diagram id="' . $DIO_DIAGRAM_ID . '" name="Page-1">
    <mxGraphModel dx="1219" dy="1005" grid="1" gridSize="10" guides="1" tooltips="1" connect="1" arrows="1" fold="1" page="1" pageScale="1" pageWidth="850" pageHeight="1100" math="0" shadow="0">
      <root>
        <mxCell id="0" />
        <mxCell id="1" parent="0" />
        <UserObject label="" plantUmlData="{&#xa;  &quot;data&quot;: &quot;' . $plantuml_escapedtextdata . '&quot;,&#xa;  &quot;format&quot;: &quot;svg&quot;&#xa;}" id="' . $DIO_USEROBJECT_ID . '">
          <mxCell style="shape=image;noLabel=1;verticalAlign=top;aspect=fixed;imageAspect=0;image=data:image/svg+xml,' . $plantuml_svgb64data . ';" parent="1" vertex="1">
            <mxGeometry x="0" y="0" width="' . $width . '" height="' . $height . '" as="geometry" />
          </mxCell>
        </UserObject>
      </root>
    </mxGraphModel>
  </diagram>
</mxfile>');

#
# Print the Draw.io file to STDOUT
#
print $drawio_xmldata;

__END__

File: plantuml_to_drawio_xml.py
#!/usr/bin/env python3

import sys
import subprocess
import base64
import xml.etree.ElementTree as ET
import os.path
import re
from time import gmtime
from datetime import datetime

JAVABIN = "java"
PLANTUMLJAR = "plantuml-1.2024.4.jar"
DEBUG = False

DIO_AGENT_NAME    = "5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.0.0 Safari/537.36"
DIO_ETAG_ID       = "ETAG111111111111xx01"
DIO_DIAGRAM_ID    = "DIAGRAM111111111xx01"
DIO_USEROBJECT_ID = "USEROBJECT111111xx01-1"

def dprint(*args, **kwargs):
    if DEBUG:
        print(*args, **kwargs)

def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <PlantUML-file>")
        sys.exit(1)

    plantuml_file = sys.argv[1]

    if not os.path.isfile(plantuml_file):
        print("Error: PlantUML file not found.")
        sys.exit(1)

    if not os.path.isfile(PLANTUMLJAR):
        print("Error: PlantUML JAR file not found.")
        sys.exit(1)

    #
    # Convert PlantUML plain text data to SVG data
    #
    # Read in PlantUML
    with open(plantuml_file, 'r') as f:
        plantuml_textdata = f.read().strip()

    dprint("PlantUML_Text=[{}]\n".format(plantuml_textdata))

    # Convert PlantUML to SVG using PlanUML Java Library
    cmd = [JAVABIN, "-jar", PLANTUMLJAR, "-tsvg", "-pipe"]
    process = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                               universal_newlines=True)

    try:
        stdout, _ = process.communicate(input=plantuml_textdata, timeout=10)
    except subprocess.TimeoutExpired:
        process.kill()
        print("Error: PlantUML process timed out.")
        sys.exit(1)

    plantuml_svgdata = stdout
    dprint("PlantUML_SVG=[{}]\n".format(plantuml_svgdata))

    #
    # base64 encode the PlantUML SVG data
    #
    plantuml_svgb64data = base64.b64encode(plantuml_svgdata.encode()).decode()
    dprint("PlantUML_Base64=[{}]\n".format(plantuml_svgb64data))

    #
    # XML/HTML/URL encode the PlantUML text data
    # Only 5 characters are needed to be escaped for XML, and include newlines and tabs
    #
    plantuml_escapedtextdata = (
        plantuml_textdata
        .replace('&', '&amp;')
        .replace('<', '&lt;')
        .replace('>', '&gt;')
        .replace('"', '&quot;')
        .replace("'", '&apos;')
        .replace('\r\n', '\\n')
        .replace('\n', '\\n')
        .replace('\r', '\\n')
        .replace('\t', '\\t')
    )
    dprint("PlantUML_EscapedText=[{}]\n".format(plantuml_escapedtextdata))

    #
    # Obtain the SVG dimensions
    #
    svg_width = ''
    svg_height = ''
    svg_dimensions_match = re.match(r'.*width:([\d]+)px;height:([\d]+)px;.*', plantuml_svgdata)
    if svg_dimensions_match:
        svg_width = svg_dimensions_match.group(1)
        svg_height = svg_dimensions_match.group(2)

    dprint("[width={},height={}]\n".format(svg_width, svg_height))

    #
    # Set the Draw.io revision date to the last modified date of the PlantUML file
    #
    file_mtime = os.path.getmtime(plantuml_file)
    file_gmtime = gmtime(file_mtime)
    gmt_timestamp = datetime.fromtimestamp(file_mtime).strftime('%Y-%m-%dT%H:%M:%S.000Z')

    dprint("[gmtime={}]\n".format(gmt_timestamp))

    #
    # Create the Draw.io file with imported PlantUML code and SVG
    # XML Data Template pulled from a random source file
    #
    # <mxfile ...>
    xml_root = ET.Element('mxfile', {
        'host': 'app.diagrams.net',
        'modified': gmt_timestamp,
        'agent': DIO_AGENT_NAME,
        'etag': DIO_ETAG_ID,
        'version': '20.0.1',
        'type': 'embed'
    })

    # <diagram ...>
    diagram_elem = ET.SubElement(xml_root, 'diagram', {'id': DIO_DIAGRAM_ID, 'name': 'Page-1'})

    # <mxGraphModel ...>
    mxGraphModel_elem = ET.SubElement(diagram_elem, 'mxGraphModel', {
        'dx': '1219', 'dy': '1005', 'grid': '1', 'gridSize': '10',
        'guides': '1', 'tooltips': '1', 'connect': '1', 'arrows': '1',
        'fold': '1', 'page': '1', 'pageScale': '1', 'pageWidth': '850',
        'pageHeight': '1100', 'math': '0', 'shadow': '0'
    })

    # <root ...>
    root_elem = ET.SubElement(mxGraphModel_elem, 'root')

    # <mxCell ...>
    mxCell_0_elem = ET.SubElement(root_elem, 'mxCell', { 'id': '0' })

    # <mxCell ...>
    mxCell_1_elem = ET.SubElement(root_elem, 'mxCell', { 'id': '1', 'parent': '0' })

    # <UserObject ...>
    userObject_elem = ET.SubElement(root_elem, 'UserObject', {
        'label': '',
        'plantUmlData': '{{\n  "data": "{}",\n  "format": "svg"\n}}'.format(plantuml_escapedtextdata),
        'id': DIO_USEROBJECT_ID
    })

    # Embed SVG image data
    svg_image_data = 'data:image/svg+xml,{}'.format(plantuml_svgb64data)
    # <mxCell ...>
    mxCell_elem = ET.SubElement(userObject_elem, 'mxCell', {
        'style': 'shape=image;noLabel=1;verticalAlign=top;aspect=fixed;imageAspect=0;image={}'.format(svg_image_data),
        'parent': '1',
        'vertex': '1'
    })

    # <mxGeometry ... />
    mxGeometry_elem = ET.SubElement(mxCell_elem, 'mxGeometry', {
        'x': '0', 'y': '0', 'width': svg_width, 'height': svg_height, 'as': 'geometry'
    })

    #
    # Print the Draw.io file to STDOUT
    #
    xml_str = ET.tostring(xml_root, encoding='unicode')
    print(xml_str)

if __name__ == "__main__":
    main()
